packages/flutter/lib/src/gestures/binding.dart

```diff
@@ -225,6 +261,47 @@ mixin GestureBinding on BindingBase implements HitTestable, HitTestDispatcher, H
       pointerSignalResolver.resolve(event);
     }
   }
+
+  void _scheduleFrameCallback() {
+    if (_frameCallbackScheduled) {
+      return;
+    }
+    final SchedulerBinding? scheduler = SchedulerBinding.instance;
+    assert(scheduler != null);
+    if (scheduler != null) {
+      _frameCallbackScheduled = true;
+      scheduler.scheduleFrameCallback((_) {
+        _frameCallbackScheduled = false;
+        _sampleTime = scheduler.currentSystemFrameTimeStamp + samplingOffset;
+        if (!locked) {
+          _flushPointerEventQueue();
+        }
+      });
+    }
+  }
+
+  // Flag to track if a frame callback has been scheduled.
+  bool _frameCallbackScheduled = false;
+
+  // Current sample time for resampling.
+  Duration _sampleTime = Duration.zero;
+
+  // Resamplers used to filter incoming pointer events when resampling
+  // is enabled.
+  final Map<int, PointerEventResampler> _resamplers =
+      <int, PointerEventResampler>{};
+
+  /// Enable pointer event resampling by setting this to `true`.
+  ///
+  /// Resampling results in smoother event processing at the cost
+  /// of some added latency. Devices with low frequency sensors or
+  /// when the frequency is not a multiple of the display frequency
+  /// (e.g., 120Hz input and 90Hz display) benefit from this.
+  bool resamplingEnabled = false;
+
+  /// Offset relative to current frame time that should be used for
+  /// resampling. `samplingOffset` is expected to be negative.
+  Duration samplingOffset = _defaultSamplingOffset;
```

Comment:
If this offset can be device refresh rate dependent, should we also calculate this number instead of set a fixed one? I was told we can get the refresh rate from a service protocol `_flutter.getDisplayRefreshRate`, if that information is somewhat helpful.

Reply:
Refresh rate is not meaningful for this offset. input sensor frequency is but that's not something that platforms typically expose.

I think we should create a follow up task to determine good default values for different devices. The best way to do that is likely through some automatic test that can also be used to detect regressions. Automatic test is how the current default value was established on Fuchsia.

Reply:
So as a particular case, Pixel4 has input frequency about 120Hz, and refresh rate 90Hz, and one use case we are planning for the resampler is to help resample this 120Hz input to 90Hz, for better scrolling smoothness. Does this sounds like need a smaller `samplingOffset`?

Reply:
samplingOffset being too large will not hurt smoothness but it will hurt input latency so it's in our interest to keep it as low as possible while still making sure resampling works well.

If input frequency is 120Hz then we need 8.333 ms to ensure that we have 2 input events that can be used for resampling. We need some margin for events to be delivered from the input driver to flutter. Let's use 4.667 ms to account for the situation when device is under heavy load. 90Hz doesn't really affect anything but I'd guess that the app [VSYNC — App reads input and generates next frame](https://source.android.com/devices/graphics/implement-vsync) (when flutter produces a frame) is relatively low on a device with that refresh rate. Let's guess 9 ms.

Given above, I'd try using samplingOffset=Duration(milliseconds: -22);  // 8.333 + 4.667 + 9